protocol/shannon/sanctioned_endpoints_store.go

```diff
@@ -183,23 +183,137 @@ func (ses *sanctionedEndpointsStore) isSanctioned(
+	}
+}
+
+// string returns the string representation of the sanction key.
+// Example: "appAddr:sessionID:supplier:endpointURL"
+func (s sanctionKey) string() string {
 	return fmt.Sprintf(
 		"%s:%s:%s:%s",
-		appAddr,
-		sessionID,
-		endpoint.supplier,
-		endpoint.url,
+		s.appAddr,
+		s.sessionID,
+		s.supplier,
+		s.endpointURL,
+	)
+}
+
+func (s sanctionKey) endpointAddr() protocol.EndpointAddr {
+	return protocol.EndpointAddr(fmt.Sprintf("%s-%s", s.supplier, s.endpointURL))
+}
+
+// --------- Sanction Details ---------
+
+// getSanctionDetails returns the sanctioned endpoints for a given service ID.
+// It provides information about:
+//   - the currently sanctioned endpoints, including the reason
+//   - counts for valid and sanctioned endpoints
+//
+// It is called by the router to allow quick information about currently sanctioned endpoints.
+func (ses *sanctionedEndpointsStore) getSanctionDetails(serviceID protocol.ServiceID) devtools.ProtocolLevelDataResponse {
+	ses.logger.Info().Msgf("Getting sanction details for service ID: %s", serviceID)
+
+	permanentSanctionDetails := make(map[protocol.EndpointAddr]devtools.SanctionedEndpoint)
+	sessionSanctionDetails := make(map[protocol.EndpointAddr]devtools.SanctionedEndpoint)
+
+	// First get permanent sanctions
+	for key, sanction := range ses.permanentSanctions {
+		sanctionServiceID := protocol.ServiceID(sanction.sessionServiceID)
+
+		// Skip sanctions for service IDS other than the provided service ID
+		if sanctionServiceID != serviceID {
+			continue
+		}
+
+		ses.processSanctionIntoDetailsMap(string(key), sanction, permanentSanctionDetails)
+	}
+
+	// Then get session sanctions
+	for key, cachedSanction := range ses.sessionSanctionsCache.Items() {
+		sanction, ok := cachedSanction.Object.(sanction)
+		if !ok {
+			ses.logger.Warn().Msg("cached sanction is not a sanction")
+			continue
+		}
+
+		sanctionServiceID := protocol.ServiceID(sanction.sessionServiceID)
+
+		// Skip sanctions for service IDs other than the provided service ID
+		if sanctionServiceID != serviceID {
+			continue
+		}
+
+		ses.processSanctionIntoDetailsMap(string(key), sanction, sessionSanctionDetails)
+	}
+
+	permanentSanctionedEndpointsCount := len(permanentSanctionDetails)
+	sessionSanctionedEndpointsCount := len(sessionSanctionDetails)
+	totalSanctionedEndpointsCount := permanentSanctionedEndpointsCount + sessionSanctionedEndpointsCount
+
+	ses.logger.Info().Msgf("Returning sanction details for service ID: %s", serviceID)
+
+	return devtools.ProtocolLevelDataResponse{
+		PermanentlySanctionedEndpoints:    permanentSanctionDetails,
+		SessionSanctionedEndpoints:        sessionSanctionDetails,
+		PermamentSanctionedEndpointsCount: permanentSanctionedEndpointsCount,
+		SessionSanctionedEndpointsCount:   sessionSanctionedEndpointsCount,
+		TotalSanctionedEndpointsCount:     totalSanctionedEndpointsCount,
+	}
+}
+
+// processSanctionIntoDetailsMap processes a sanction and adds it to the provided details map
```

Comment:
#PUC what both of these actions entail.

I had to go read the code to understand.